### PR TITLE
ENH: Implement `reshape` function

### PR DESCRIPTION
**Pull request by mtsokol**
Hi @hameerabbasi,

This PR adds Array API [reshape](https://data-apis.org/array-api/latest/API_specification/generated/array_api.reshape.html) function (without `copy` keyword for now).

Unfortunately calling `sparse.reshape(...)` gives a segmentation fault (on exit) - I haven't figured out it yet.

### CODE REVIEW

**Comment by mtsokol** [comment]
@hameerabbasi The PR is ready from my side. For now `reshape` only works on CSF and CSR formats. For other ones it's broken (there are skips in the test with relevant links pointing to issues).

File: sparse/mlir_backend/_ops.py [hunk 1]
```diff
@@ -88,3 +115,25 @@ def add(x1: Tensor, x2: Tensor) -> Tensor:
...
+
+def reshape(x: Tensor, /, shape: tuple[int, ...]) -> Tensor:
+    ret_obj = x._format_class()
+    x_tensor_type = x._obj.get_tensor_definition(x.shape)
+    out_tensor_type = x._obj.get_tensor_definition(shape)
+
+    with ir.Location.unknown(ctx):
+        shape_tensor_type = ir.RankedTensorType.get([len(shape)], Index.get_mlir_type())
+
+    # TODO: Add proper caching
```
Comment (hameerabbasi) [inline sparse/mlir_backend/_ops.py:128]:
```suggestion
```

File: sparse/mlir_backend/_ops.py [hunk 2]
```diff
@@ -68,6 +70,31 @@ def add(a, b):
     return mlir.execution_engine.ExecutionEngine(module, opt_level=2, shared_libs=[MLIR_C_RUNNER_UTILS])
 
 
+# @fn_cache
```
Comment (hameerabbasi) [inline sparse/mlir_backend/_ops.py:73]:
```suggestion
@fn_cache
```

File: sparse/mlir_backend/_ops.py (hunk 1 above)
Comment (mtsokol) [reply sparse/mlir_backend/_ops.py]: Done!

File: sparse/mlir_backend/_ops.py (hunk 2 above)
Comment (mtsokol) [reply sparse/mlir_backend/_ops.py]: Done!

**Review by hameerabbasi — APPROVED** [review]

**Pull request merged**